helm-chart/splunk-kubernetes-logging/templates/_helpers.tpl

```diff
@@ -133,8 +124,135 @@ Create the name of the service account to use
+  {{- $str := $str | replace "(" "\\(" -}}
+  {{- $str := $str | replace ")" "\\)" -}}
+  {{- $str := $str | replace "[" "\\[" -}}
+  {{- $str := $str | replace "]" "\\]" -}}
+  {{- $str := $str | replace "^" "\\^" -}}
+  {{- $str := $str | replace "$" "\\$" -}}
+  {{- print "^" $str "$" -}}
+{{- end }}
+
+{{/*
+Generates the jq_transformer query to ensure each record has the appropriate index.
+*/}}
+{{- define "splunk-kubernetes-logging.index_jq_filter" -}}
+  {{- if (or .Values.splunk.hec.indexRouting .Values.global.splunk.hec.indexRouting) -}}
+    {{- include "splunk-kubernetes-logging.index_jq_filter_implicit" . -}}
+  {{- else -}}
+    {{- include "splunk-kubernetes-logging.index_jq_filter_explicit" . -}}
+  {{- end -}}
+{{- end -}}
+
+{{/*
+Generates the jq_transformer query to ensure each record has the appropriate index when routing by namespace.
+*/}}
+{{- define "splunk-kubernetes-logging.index_jq_filter_implicit" -}}
+  {{- $splunkIndexRoutingDefault := or .Values.splunk.hec.indexRoutingDefaultIndex .Values.global.splunk.hec.indexRoutingDefaultIndex "main" -}}
+  {{- print ".record | .index = (if .namespace == \"default\" then " ($splunkIndexRoutingDefault | quote) " else .namespace//" ($splunkIndexRoutingDefault | quote) " end)" -}}
+{{- end -}}
+
+{{/*
+Generates the jq_transformer query to ensure each record has the appropriate index when not routing by namespace.
+*/}}
+{{- define "splunk-kubernetes-logging.index_jq_filter_explicit" -}}
+  {{- $journalPath := or .Values.journalLogPath .Values.global.journalLogPath "/run/log/journal/" -}}
+  {{- $splunkHecIndex := (or .Values.splunk.hec.indexName .Values.global.splunk.hec.indexName) -}}
+  {{- $logDefsByIndexTag := dict -}}
+  {{- range $name, $logDef := .Values.logs -}}
+    {{- if (hasKey $logDef "index") -}}
+      {{- $index := or $logDef.index $splunkHecIndex -}}
+      {{- if (ne $index $splunkHecIndex) -}}
+        {{- $sourcetype := or $logDef.sourcetype $name -}}
+        {{- $tag := dict "value" "" -}}
+
+        {{- if $logDef.from.pod -}}
+          {{- $_ := set $tag "value" "tail.containers.*" -}}
+        {{- else if $logDef.from.file -}}
+          {{- $_ := set $tag "value" (print "tail.file." $sourcetype) -}}
+        {{- else if $logDef.from.journald -}}
+          {{- $_ := set $tag "value" (print "journald." $sourcetype) -}}
+        {{- else -}}
+          {{- print "def TODO_support_log_" $name ": " ($logDef | toJson) "; " -}}
+        {{- end }}
+
+        {{- $logDefsByTag := or (pluck $index $logDefsByIndexTag | first) (dict) -}}
+        {{- $logDefs := or (pluck $tag.value | first) (list) -}}
+        {{- $appendedLogDefs := append $logDefs (and (set $logDef "name" $name)) -}}
+        {{- $appendedLogDefsByTag := set $logDefsByTag $tag.value $appendedLogDefs -}}
+        {{- $_ := set $logDefsByIndexTag $index $appendedLogDefsByTag -}}
+      {{- end }}
+    {{- end }}
+  {{- end }}
+
+  {{- $isFirstIndex := dict "value" true -}}
```

Comment:
I am unclear on the implementation details for $isFirstIndex, could you please elaborate more.

Reply:
`$isFirstIndex` is just a variable we define here, and we set it to a dictionary with a single property `value` and the value of that property is `true`. The reason we use a dictionary is that we can reassign the properties of a dictionary, but we can't reassign boolean variables directly (any attempt to do so will create a new shadow variable that exists only within local scope, once that scope is left we would revert to the previous variable).

The purpose of the `$isFirstIndex` variable is to provide different output for the first item in the range than we do for the other items. (`if` as opposed to `elif`)